features/253-auth-sdk-callback-in-next/api.md

```diff
@@ -0,0 +1,324 @@
+delete currentUser['metadata']['username'];
+skygear.auth.updateMetadata(currentUser);
+// user has following two login IDs:
+// - "example@example.com"(email) + "password"(password)
+// - "john.doe"(nickname) + "john.doe@example.com"(nickname) + "password"(password)
+// user object's metadata is
+// {
+//   //------ `loginIDs`:
+//   "email": "example@example.com",
+//   "nickname": "john.doe",
+//   "business_email": "john.doe@example.com",
+//   //------ user custom attributes:
+//   "gender": "none"
+// }
+```
+
+# `updateMetadata` and user custom attributes
+
+`updateMetadata` also supports to update user's custom attributes, any data in `metadata` does not belong to `loginIDs` is saved properly when the user invokes `updateMetadata`.
+
+# Changes on Client JS SDK
+
+Because user object is not a `record` object anymore, client SDK has to treat user object as a plain object, and a user can access user attributes via regular way.
+
+In skygear v1 JS SDK, `container`'s `UserRecord` is defined as
+
+```
+export const UserRecord = Record.extend('user');
+```
+
+Since `Record` is removed, `UserRecord` in APIs should be replaced as `User` class (a simple plain JavaScript class), they are:
+
+```
+- currentUser: User
+- async changePassword(oldPassword: String, newPassword: String, invalidate: Boolean): Promise<User>
+- async fetchUserRole(users: User[] | String[]): Promise<Object>
+- async login(loginIDs: Object, password: String): Promise<User>
+- async loginWithEmail(email: String, password: String): Promise<User>
+- async loginWithProvider(provider: String, loginIDs: Object): Promise<User>
+- async loginWithUsername(username: String, password: String): Promise<User>
+- async signup(loginIDs: Object, password: String, data: Object): Promise<User>
+- async signupAnonymously(): Promise<User>
+- async signupWithEmail(email: String, password: String, data: Object): Promise<User>
+- async signupWithUsername(username: String, password: String, data: Object): Promise<User>
+- async whoami(): Promise<User>
+```
+
+## New API on Client JS SDK
+
+- `Promise<User> updateMetadata(<User>);`  
+  use for updating a user's metadata.
+
+## Removal APIs on Client JS SDK
+
+Due to the removal of record gear, auth gear and client SDK won't provide below functionalities:
+
+1. user query
+2. admin related API (disable user, change role, ...)
+
+And so below APIs are removed from client SDK:
+
+- `adminDisableUser`
+- `adminEnableUser`
+- `adminResetPassword`
+- `assignUserRole`
+- `revokeUserRole`
+- `setDefaultRole`
+- `setAdminRole`
+- `onUserChanged`
+
+Those requirements can be implemented via cloud function or external user DB.
+
+## API argument naming chages on Client JS SDK:
+
+- `public async requestVerification(recordKey: String): Promise`  
+  `public async requestVerification(loginIDKeys: [String]): Promise`
+- `async login(authData: Object, password: String): Promise<Record>`  
+  `async login(loginIDs: Object, password: String): Promise<User>`
```

Comment:
Shall we have an API for login with LoginID without specifiying which field to match? Something like:
`async loginWithLoginID(loginID: String, password: String)`?

Reply:
The usage of `async login(loginIDs: Object, password: String): Promise<User>` is
```
skygear.auth.login({username: "exmaple"}, "123456");
```
auth gear will match `{username: "example"}` with existing principals to authenticate the user and the password.

In your suggestion API `async loginWithLoginID(loginID: String, password: String)`? Do you expect user login by `loginWithLoginID("example", "123456")`? User doesn't need to specifify  `username` here.

Reply:
> In your suggestion API `async loginWithLoginID(loginID: String, password: String)`? Do you expect user login by `loginWithLoginID("example", "123456")`? User doesn't need to specifify `username` here.

yes, but maybe just ignore me here first lol

Reply:
Thanks, I ignore this one temporarily.